### DISCUSSION TITLE
Really long type of `ServiceBuilder`

### DISCUSSION THREAD

**Discussion by dulguun0225** [tokio-rs/axum#1855] [Q&A] [answered]
I wanted to put creation of layers into one function and pass it to `Router::layer`. But I can't figure out what the return type should be.

VSCode tells me that type of `let builder:` is `ServiceBuilder<Stack<TimeoutLayer, Stack<HandleErrorLayer<fn handle_timeout_error(Box<dyn Error + Send + Sync>) -> impl Future<Output = (StatusCode, String)>, {unknown}>, Stack<CompressionLayer, Stack<TraceLayer<SharedClassifier<ServerErrorsAsFailures>>, Identity>>>>>`

But I'm guessing it should be something simpler on the return type.

```rust
pub fn new() -> ??????????????? {
    let builder = ServiceBuilder::new()
        .layer(TraceLayer::new_for_http())
        .layer(CompressionLayer::new())
        .layer(HandleErrorLayer::new(handle_timeout_error))
        .timeout(Duration::from_secs(30));

    builder
}
```

**Comment by davidpdrsn** [comment, marked as answer]
That's just how tower works. Nothing axum do about that.

I general I recommend writing

```rust
fn apply_middleware(router: Router) -> Router {
    ...
}
```

instead.
